### ISSUE TITLE
FA issue

### ISSUE THREAD

**Issue by arpruss** [SvOlli/retrocartdumper#4]
Looking at your code, unless I missed something, you don't assert the D0 line when accessing the hotspots for the FA mapper. According to the patent for the FA mapper, the D0 line needs to be asserted when accessing the hotspots. (The patent says something about how this fixes some timing issue. I wonder if it isn't mainly just to get around an Atari patent or something like that.)

**Comment by SvOlli** [comment]
This is correct. I just took the knowledge from Stella / Uno Cart to learn what to trigger. Do you have a link to the patent, so I can learn on what I should do?

**Comment by dionoid** [comment]
I think this is the patent document being referred to:
https://forums.atariage.com/applications/core/interface/file/attachment.php?id=510935

**Comment by dionoid** [comment]
(removed my comment, as I thought this was about Activision's FE bank switching, not CBS' FA)

**Comment by arpruss** [comment]
I also have a bug report to the Stella people, and it looks like they're going to fix it.

You'll get a link to the patent and the discussion of my bug report here:
https://github.com/stella-emu/stella/issues/987

**Comment by SvOlli** [comment]
Ah, if it's about pulling D0 high, I'm doing this. Take a look at https://github.com/SvOlli/retrocartdumper/blob/74b1520655b17c527dbd255a353ef30a4171b9d9/atari2600/dump_fx.c#L27 :
poke( serfd, 0xF000, 0xFF ); /* neat trick to clean up superchip ram during dump */

This sets the address bus to $1000 and the data bus to $FF before changing the address bus to read any ROM data. I'm originally doing this, so if I trigger the SARA RAM chip during the read it now gets written and is supplying me with a nice $FF.

As a part of this it also sets D0 to 1 as required by FA bankswitching, if I understood you correctly.

**Comment by SvOlli** [comment]
Damn, an as I see now, I'm only doing this before reading SARA RAM area. But it should be easy to add to make the dumper more compliant.

**Comment by arpruss** [comment]
I also have to confess that I don't know how that one poke cleans up all of the RAM area. Isn't the write port 0x1000-0x107F, so to clean it all up, you'd need to write 0xFF to each byte of the write port?

If you're using a 5V device with an internal pullup, can't you maybe solve the whole problem by turning on internal pullups for input? You should be able to then clean up the RAM area just by reading the write port, and you automatically have D0 asserted when you need it?

(In my own 2600 dumper, I can't do that elegant solution, because I'm using a 3.3V microcontroller, and even though I am using nominally 5V tolerant pins, according to the datasheet the pins are no longer 5V tolerant when you turn on the internal pull up/down resistors. I also have a lot more complexity because I try to auto-detect the mapping scheme to make it easier for the user: https://github.com/arpruss/atari2600dumper .)

**Comment by thrust26** [comment]
I have added some comments/questions to the Stella [issue](https://github.com/stella-emu/stella/issues/987). Maybe one of you can help?
